### PR TITLE
Add apply_local_tensor() method to torch.nn.Module.

### PR DESCRIPTION
**Pull request by jma127**
This commit adds the ``apply_local_tensor()`` method to torch.nn.Module.

``apply_local_tensor()`` is a nonrecursive version of the ``_apply()`` private method.

See https://github.com/pytorch/ELF/pull/78 for an example where the
non-recursive variant is needed.

### CODE REVIEW

**Comment by jma127** [comment]
There are no unit tests for the old ``_apply()`` (or public-facing ``apply()``, for that matter). So I omit testing for now.

Bugs should be caught by the various tests for ``half()``, etc (as well as my manual testing with ELF code).

**Comment by jma127** [comment]
Can someone help me figure out this lint error? I can't see what it has to do with this PR.

https://travis-ci.org/pytorch/pytorch/jobs/414765331

**Comment by jma127** [comment]
Checked in w/ @MisterTea -- it seems like 31646edfff2678b0149a673ef5160a23c77e3cb9 introduced a lint error. FB's internal repo recently switched to Black, and it seems that its formatting is inconsistent with flake8 here.

@MisterTea will address it in a few days.

**Comment by ezyang** [comment]
#10441

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -241,6 +223,53 @@ def apply(self, fn):
...
 
+    def apply_data(self, fn, to_parameters=True, to_gradients=True, to_buffers=True, recursive=True):
+        r"""Applies ``fn`` to the various tensors within the module.
+
+        This method differs from ``apply()`` in that it acts on the tensors of
+        the module rather than the module itself.
+
+        Args:
+            fn (:class:`Tensor` -> None): function to be applied to each tensor.
+            to_parameters (bool): whether to apply the function to parameters.
```
[the remark this answers is not in the document]
Comment (jma127) [reply torch/nn/modules/module.py]: Thanks for flagging, will fix.

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -241,6 +223,53 @@ def apply(self, fn):
...
+        """
+        if recursive:
+            for module in self.children():
+                module.apply_data(fn)
+
+        for param in self._parameters.values():
+            if param is not None and to_parameters:
+                # Tensors stored in modules are graph leaves, and we don't
+                # want to create copy nodes, so we have to unpack the data.
+                param.data = fn(param.data)
```
[the remark this answers is not in the document]
Comment (jma127) [reply torch/nn/modules/module.py]: Thanks for flagging, will fix.

**Comment by jma127** [comment]
Addressed review comments.

**Comment by jma127** [comment]
Rebased to get @ezyang 's lint fix in. @SsnL can you take another look when you get a chance?

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -241,6 +223,69 @@ def apply(self, fn):
...
+                whether to apply the function to buffers.
+            recursive (bool):
+                whether to apply the function to tensors in
+                all submodules.
+
+        Returns:
+            Module: self
+
+        Example:
+            # Converts model's non-batchnorm parameters to half precision.
```
Comment (SsnL) [inline torch/nn/modules/module.py]: The example format should be consistent with other examples we have, e.g., prefix with >>> and others. See https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L214 for an example.

**Review by ssnl — COMMENTED** [review]
I'm still a bit concerned with the motivation for this change. I can't think of much use other than solving the BN .half() issue. And this doesn't deal with that very cleanly. Can we instead rewrite the conversion methods so that subclasses can overwrite them, and overwrite in _BatchNorm?

**Comment by jma127** [comment]
@SsnL I don't think conversion behavior should be a part of the module definition. Some users may be happy (and in fact, may prefer) 16b-quantized floats, some users may want the original floats, etc. Rather than enforcing one-off conversion behavior for specific modules, I think we should make the default behavior consistent and predictable (i.e. ``half()`` will give you a fully fp16 module no matter what). If you want anything fancier, you can be explicit by manually iterating through the individual submodules and specifying exactly what conversion you want to perform (as in the provided example).

**Comment by ezyang** [comment]
I'm not sure if the design is right or not, but this certainly needs a test. cc @apaszke

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -241,6 +223,69 @@ def apply(self, fn):
...
+                # Tensors stored in modules are graph leaves, and we don't
+                # want to create copy nodes, so we have to unpack the data.
+                param.data = fn(param.data)
+            if param is not None and param._grad is not None and to_gradients:
+                # Tensors stored in modules are graph leaves, and we don't
+                # want to create copy nodes, so we have to unpack the data.
+                param._grad.data = fn(param._grad.data)
+
+        for key, buf in self._buffers.items():
+            if buf is not None and to_buffers:
```
Comment (ezyang) [inline torch/nn/modules/module.py]: Hoist the `to_buffers` check outside of the loop

**Comment by SsnL** [comment]
@jma127 Sure. That makes sense. But I'm still not quite convinced that we should have this. It just doesn't feel a good solution to the problem you had with fp16, and I can't see much use other than that.

Alternatively, how about exposing methods to iterate through immediate parameters & buffers, e.g., via a flag on existing methods? Then users can just use `apply` and it will be much more general.

File: torch/nn/modules/module.py [hunk 5]
```diff
@@ -241,6 +223,69 @@ def apply(self, fn):
         fn(self)
         return self
 
+    def apply_tensor(self, fn, to_parameters=True, to_gradients=True,
+                     to_buffers=True, recursive=True):
```
Comment (ezyang) [inline torch/nn/modules/module.py]: Is the generality of the `to_` arguments needed? If you don't need them for your use case, I'd rather drop them for now and add them later if someone shouts.

I'm also troubled as to whether or not `apply_tensor` should be nonrecursive by default. You can always spell the recursive version as `mod.apply(lambda m: m.apply_tensor(fn))`

Comment (jma127) [reply torch/nn/modules/module.py]: Both good points -- for the second, I would also like users to be explicit by calling apply. I'll make these changes.

**Comment by jma127** [comment]
@SsnL I considered that exact thing (i.e. some ``iter_tensors`` method). The problem I envisioned was that you can't perform certain operations on tensors in-place (certainly not something like ``half()``).

The additional functionality that ``apply_tensor`` provides over something like ``iter_tensor`` is to support these non-inplace operations.

File: torch/nn/modules/rnn.py [hunk 6]
```diff
@@ -108,11 +108,16 @@ def flatten_parameters(self):
...
-    def _apply(self, fn):
-        ret = super(RNNBase, self)._apply(fn)
+    def apply_tensor(self, fn, to_parameters=True, to_gradients=True, to_buffers=True, recursive=True):
+        ret = super(RNNBase, self).apply_tensor(
+            fn=fn, to_parameters=to_parameters, to_gradients=to_gradients,
+            to_buffers=to_buffers, recursive=recursive)
         self.flatten_parameters()
         return ret
 
+    def _apply(self, fn):
```
[the remark this answers is not in the document]
Comment (jma127) [reply torch/nn/modules/rnn.py]: Yep good catch ;)

**Comment by jma127** [comment]
Also, to explain my usecase as clearly as possible:

Current recommended methods of converting batchnorm models to FP16 in pytorch all involve calling ``half()`` on the entire module, then calling ``float()`` on individual BatchNorm layers. This is undesirable because the batchnorm parameters undergo an unnecessary conversion from fp32 to fp16 back to fp32. Besides the reduced precision that this induces, this can actually cause NaNs/infs in what would otherwise be a perfectly good model (if any of the batchnorm parameters are above fp16's max value).

My motivation for this PR is that users should not have to touch private variables (``_parameters`` and ``_buffers``) in order to do either this or any other module-specialized tensor operations.

**Comment by SsnL** [comment]
@jma127 You still can do operations inplace, e.g.,
```
def fn(m):
    for buffer in m.buffers(immediate_child=True):
        buffer.data = torch.randn(3)
```

**Comment by jma127** [comment]
Ah, any idea why the original code for ``_apply()`` did it like this then?

    for key, buf in self._buffers.items():
        if buf is not None:
            self._buffers[key] = fn(buf)

**Comment by SsnL** [comment]
@jma127 Probably for safety reasons, fiddling with .data is unsafe in many ways, e.g., here it keeps the object identity but changes the content completely.

Alternatively, a safer option can be done with named_parameters, i.e.,

```
def fn(m):
    for name, buffer in m.named_buffers(immediate_child=True):
        if isinstance(m, _BatchNorm) and (name == 'running_mean' or name == 'running_var'):
            m.name = buffer.float()
        else:
            m.name = buffer.half()
    ....

net.apply(fn)
```

**Review by apaszke — CHANGES_REQUESTED** [review]
I don't think this function is a good addition to our API. I think a better solution would be to get `named_buffers()` (to match `named_parameters()`). Then, you can write this helper in your codebase if you really need it:
```python
def my_local_apply(module, f):
    named_tensors = itertools.chain(module.named_parameters(), module.named_buffers())
    for n, t in named_tensors:
        setattr(module, n, f(t))
```

**Comment by jma127** [comment]
@apaszke : this could be workable. How would you suggest to convert the gradient tensors (for parameters)? Or is it not necessary to convert those because of the "parent" tensor's conversion?

**Comment by apaszke** [comment]
I'm not 100% sure, but gradients of parameters are in their `.grad` attributes, and so if you cast/replace a parameter, you can easily replace those too.

**Comment by jma127** [comment]
OK. This solution would require the addition of the immediate_child flag, like @SsnL suggested. Is that acceptable to have in all these named_* methods?

**Comment by apaszke** [comment]
Ohhh, thanks. `immediate_child` is very confusing since generally children are the submodules of the module, so "a child" is not a well-defined concept I'd say. How about adding `recurse=False` to `(named_)parameters` and `(named_)buffers`?

**Comment by jma127** [comment]
Great, this is what I had in mind for the original function (add_local_tensors). I'll close this PR and create a new one with these changes.

**Comment by jma127** [comment]
Successor PR at #10554
